### ISSUE TITLE
[RFC] Add argument to specify output nonlinearity

### ISSUE THREAD

**Issue by BenjaminBossan** [skorch-dev/skorch#661]
Currently, the nonlinearity applied to the module output before returning it from `predict_proba` is not changeable.

For `NeuralNetClassifier`, no nonlinearity is applied, which means that if the module doesn't return probabilities (which might be desired, e.g. for numerical stability), `predict_proba` also doesn't return probabilities.

For `NeuralNetBinaryClassifier`, `sigmoid` is applied if the loss is an instance of `torch.nn.BCEWithLogitsLoss`. Otherwise, it's untouched.

For `NeuralNet` and `NeuralNetRegressor`, no nonlinearity is applied, which is typically what is desired (we still provide `predict_proba`, even though it's just the same as `predict`, but it might still be useful for users to have two methods).

The proposal is to add an argument to `NeuralNet` called `output_nonlinearity` (or similar) that is applied to the module's result. This makes it very easy for the user to change the output nonlinearity.

Where is the right place to apply it? I would argue on a batch level inside `forward_iter`, before calling `to_device`. There are three main reasons:

1. Use GPU acceleration if so desired
2. Use pytorch functions instead of numpy -- it's more likely that pytorch provides an appropriate function
3. Not only affects `predict_proba`, but also `forward` and `forward_iter`, which should be preferable for most

Regarding the implementation, it should be backwards compatibly, i.e. the solution should do exactly the same as is being done at the moment. To achieve this, the default `output_nonlinearity` should be `'auto'`. In that case, a special function should be called that infers the nonlinearity from the `net` instance and reproduces the existing behavior. We could think about extending the inferred nonlinearities compared to what is done right now, e.g. to apply `softmax` is the loss is `CrossEntropyLoss`.

(An interesting design decision would be if the `auto` function "dispatches" only on the `criterion` or also on the `net`. E.g., if the `criterion` is `BCEWithLogitsLoss`, is `sigmoid` always applied, or only in case of `NeuralNetBinaryClassifier`? Regardless of the answer, I would like to have the `auto` function take the whole net as argument, not only the the criterion, to infer the nonlinearity, even if we decide to only use the criterion for now, in case we need/want to extend it further.)

Other options for `output_nonlinearity` should be callables, like `torch.softmax`, or `None`, in which case the output is untouched.

This proposal supersedes #572 and #580 (which seem to be abandoned anyway).

Note that this change _might_ break existing code, where methods like `predict_proba` are overridden and make the assumption that the outputs from `forward_iter` are untouched. Those users should set `output_nonlinearity=None` to make their code work again. I think this is an acceptable tradeoff (document this in the CHANGES.md).

**Comment by thomasjpfan** [comment]
I agree with adding `output_nonlinearity='auto'`, because it is more explicit. Would be the other accepted values for `output_nonlinearity` by pytorch functions or  any callable?

> An interesting design decision would be if the auto function "dispatches" only on the criterion or also on the net

Since we are currently dispatch based on criterion, we can continue doing that for now. I am assuming that the auto function will be private that takes the whole net.

**Comment by BenjaminBossan** [comment]
> Would be the other accepted values for `output_nonlinearity` by pytorch functions or any callable?

I think we could accept any callable, it would be hard to check (or even define) what a "pytorch function" would be.

> Since we are currently dispatch based on criterion, we can continue doing that for now.

I agree, we can always move to more fine grained dispatching later.

> I am assuming that the auto function will be private that takes the whole net.

Yes, I think it should take the net as input.

As to how the auto function is implemented, I'm still not quite sure. E.g. it could make use of a "dispatching table" (which at the moment doesn't differ for the net class, only for the criterion) with the possibility for the user to enhance it by just modifying it. But maybe that's overengineered for this problem :D
